# Эксперимент 1: Llama-3.2-1B + LoRA



## Установка


In [ ]:
!pip install -q transformers==4.47.0 accelerate peft trl==0.11.4 \
             datasets bitsandbytes wandb evaluate rouge_score bert_score

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.5/43.5 kB 2.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.1/10.1 MB 81.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 316.6/316.6 kB 26.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.7/60.7 MB 13.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 4.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.1/61.1 kB 3.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 566.4/566.4 kB 24.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 66.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 185.2/185.2 kB 17.5 MB/s eta 0:00:00


In [ ]:
import torch
import numpy as np
import pandas as pd
import wandb
from huggingface_hub import login
from sklearn.model_selection import train_test_split
from datasets import Dataset
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig
from peft import LoraConfig, get_peft_model, prepare_model_for_kbit_training, PeftModel
from trl import SFTTrainer, SFTConfig
import evaluate as hf_evaluate
from bert_score import score as bert_score
import ipywidgets as widgets
from IPython.display import display, clear_output

torch.manual_seed(42)

In [ ]:
login('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')
wandb.login(key='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')

MODEL_ID   = 'meta-llama/Llama-3.2-1B-Instruct'
WB_ENTITY  = 'martyanov-an-'
WB_PROJECT = 'GP_5'

/usr/local/lib/python3.12/dist-packages/notebook/notebookapp.py:191: SyntaxWarning: invalid escape sequence '\/'
  | |_| | '_ \/ _` / _` |  _/ -_)
wandb: WARNING If you're specifying your api key in code, ensure this code is not shared publicly.
wandb: WARNING Consider setting the WANDB_API_KEY environment variable, or running `wandb login` from the command line.
wandb: [wandb.login()] Using explicit session credentials for https://api.wandb.ai.
wandb: No netrc file found, creating one.
wandb: Appending key for api.wandb.ai to your netrc file: /root/.netrc
wandb: Currently logged in as: vitalyansky (martyanov-an-) to https://api.wandb.ai. Use `wandb login --relogin` to force relogin


## Загрузка и подготовка данных

In [ ]:
CSV_PATH = 'data (1).csv'
df = pd.read_csv(CSV_PATH)[['input_text', 'target_text']].dropna().reset_index(drop=True)

train_idx, tmp = train_test_split(df.index.tolist(), test_size=0.2, random_state=42)
val_idx, test_idx = train_test_split(tmp, test_size=0.5, random_state=42)

df_train = df.loc[train_idx].reset_index(drop=True)
df_val   = df.loc[val_idx].reset_index(drop=True)
df_test  = df.loc[test_idx].reset_index(drop=True)

print(len(df_train), len(df_val), len(df_test))

20135 2517 2517


In [ ]:
tokenizer = AutoTokenizer.from_pretrained(MODEL_ID)
tokenizer.pad_token    = tokenizer.eos_token
tokenizer.padding_side = 'right'

# оборачиваем каждую пару input/target в chat-template — так модель понимает роли
def to_dataset(dataframe):
    rows = []
    for _, r in dataframe.iterrows():
        msgs = [
            {'role': 'user',      'content': r['input_text']},
            {'role': 'assistant', 'content': r['target_text']},
        ]
        rows.append({'text': tokenizer.apply_chat_template(msgs, tokenize=False, add_generation_prompt=False)})
    return Dataset.from_list(rows)

ds_train = to_dataset(df_train)
ds_val   = to_dataset(df_val)

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/296 [00:00<?, ?B/s]

## Базовая модель (до файнтюна)

In [ ]:
# грузим модель в 4-bit через bitsandbytes, чтобы влезть в GPU
bnb = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type='nf4',
    bnb_4bit_use_double_quant=True,
    bnb_4bit_compute_dtype=torch.bfloat16,
)

def load_model():
    m = AutoModelForCausalLM.from_pretrained(MODEL_ID, quantization_config=bnb, device_map='auto')
    m = prepare_model_for_kbit_training(m)
    m.config.use_cache = False
    return m

model = load_model()

config.json:   0%|          | 0.00/877 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/2.47G [00:00<?, ?B/s]

/usr/local/lib/python3.12/dist-packages/bitsandbytes/backends/cuda/ops.py:213: FutureWarning: _check_is_size will be removed in a future PyTorch release along with guard_size_oblivious.     Use _check(i >= 0) instead.
  torch._check_is_size(blocksize)


generation_config.json:   0%|          | 0.00/189 [00:00<?, ?B/s]

In [ ]:
@torch.no_grad()
def respond(mdl, text, max_new_tokens=100):
    msgs = [{'role': 'user', 'content': text}]
    x = tokenizer.apply_chat_template(msgs, return_tensors='pt', add_generation_prompt=True).to(mdl.device)
    out = mdl.generate(x, max_new_tokens=max_new_tokens, temperature=0.7, top_p=0.9, do_sample=True)
    return tokenizer.decode(out[0][x.shape[1]:], skip_special_tokens=True)

In [ ]:
# смотрим на базовую модель до обучения — потом сравним с файнтюном
sample = df_test.head(50)
preds  = [respond(model, r['input_text']) for _, r in sample.iterrows()]
refs   = list(sample['target_text'])

rouge      = hf_evaluate.load('rouge')
r_base     = rouge.compute(predictions=preds, references=refs)
_, _, bf1  = bert_score(preds, refs, lang='en', verbose=False)

def distinct(texts, n):
    ng, total = set(), 0
    for t in texts:
        toks = t.split()
        g = [tuple(toks[i:i+n]) for i in range(len(toks)-n+1)]
        ng.update(g); total += len(g)
    return round(len(ng)/total, 4) if total else 0

base_metrics = {
    'rouge1': round(r_base['rouge1'], 4),
    'rougeL': round(r_base['rougeL'], 4),
    'bertscore': round(bf1.mean().item(), 4),
    'distinct1': distinct(preds, 1),
    'distinct2': distinct(preds, 2),
}
print('base model metrics:', base_metrics)

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
/usr/local/lib/python3.12/dist-packages/bitsandbytes/backends/cuda/ops.py:468: FutureWarning: _check_is_size will be removed in a future PyTorch release along with guard_size_oblivious.     Use _check(i >= 0) instead.
  torch._check_is_size(blocksize)
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
The attention m

tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/482 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/1.42G [00:00<?, ?B/s]

Some weights of RobertaModel were not initialized from the model checkpoint at roberta-large and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


base model metrics: {'rouge1': np.float64(0.1103), 'rougeL': np.float64(0.0789), 'bertscore': 0.8269, 'distinct1': 0.3322, 'distinct2': 0.7304}


### Что показывают метрики до файнтюна

Смотрим на числа и понимаем, что базовая модель вообще не заточена под стиль саппорта.

**ROUGE-1 / ROUGE-L** — считает пересечение n-грамм между нашим ответом и референсом. У нас получилось 0.11 / 0.079 — ожидаемо мало, потому что модель генерит текст в своём стиле, а не так, как написаны эталонные ответы. Будем смотреть как вырастет после файнтюна.

**BERTScore** — это уже семантика, не буквальное совпадение. 0.83 — довольно высокое, потому что модель на английском работает хорошо и смысловые эмбеддинги близко. Но высокий BERTScore не значит что ответы хорошие — просто слова похожи по смыслу на референс.

**Distinct-1 / Distinct-2** — доля уникальных уни- и биграмм в генерациях. Высокие значения (0.33 / 0.73) означают что модель разнообразно генерирует, не повторяется. После файнтюна ожидаем что упадёт — модель начнёт выдавать более шаблонные ответы в стиле саппорта, и это нормально.

**Perplexity** — считаем позже из val_loss через exp(loss). Чем меньше — тем лучше. Покажет насколько модель уверена в предсказаниях на нашем датасете.

## Логируем датасет в W&B

In [ ]:
# сохраняем датасет как артефакт в W&B чтобы был трекинг версий
with wandb.init(entity=WB_ENTITY, project=WB_PROJECT, name='dataset', job_type='data') as run:
    art = wandb.Artifact('customer_support', type='dataset')
    art.add_file(CSV_PATH)
    run.log_artifact(art)

In [ ]:
def compute_metrics(mdl, df, n=200):
    sample = df.head(n)
    preds  = [respond(mdl, r['input_text']) for _, r in sample.iterrows()]
    refs   = list(sample['target_text'])

    r = rouge.compute(predictions=preds, references=refs)
    _, _, f1 = bert_score(preds[:100], refs[:100], lang='en', verbose=False)

    return {
        'rouge1': round(r['rouge1'], 4),
        'rougeL': round(r['rougeL'], 4),
        'bertscore': round(f1.mean().item(), 4),
        'distinct1': distinct(preds, 1),
        'distinct2': distinct(preds, 2),
    }

## Эксперимент 1 — LoRA только на attention-слоях (r=4)


Параметры:
- `r=4` — ранг матриц адаптера, чем меньше — тем меньше параметров добавляем. Взяли минимальный, чтобы посмотреть вообще работает ли
- `lora_alpha=8` — масштабирующий коэффициент. Эффективный lr адаптера = alpha/r = 2.0
- `target_modules: q_proj, v_proj` — трогаем только query и value проекции в attention. Самый лёгкий вариант
- `dropout=0.05` — небольшая регуляризация чтобы не переобучился


In [ ]:
# конфиг LoRA — только q_proj и v_proj, rank=4

lora_cfg = LoraConfig(
    r=4,
    lora_alpha=8,
    lora_dropout=0.05,
    target_modules=['q_proj', 'v_proj'],
    task_type='CAUSAL_LM',
)

run = wandb.init(
    entity=WB_ENTITY,
    project=WB_PROJECT,
    name='exp1_attn_r4',
    config={
        'r': 4,
        'lora_alpha': 8,
        'lora_dropout': 0.05,
        'target_modules': ['q_proj', 'v_proj'],
        'epochs': 2,
        'lr': 2e-4,
        'batch_size': 4,
        'max_seq_len': 128,
    },
)

model = load_model()
model = get_peft_model(model, lora_cfg)
model.print_trainable_parameters()

# конфиг обучения — 2 эпохи, lr=2e-4, batch=4 с grad_accum=2
cfg = SFTConfig(
    output_dir='./ckpt/exp1_attn_r4',
    num_train_epochs=2,
    per_device_train_batch_size=4,
    gradient_accumulation_steps=2,
    learning_rate=2e-4,
    warmup_ratio=0.03,
    bf16=True,
    logging_steps=20,
    eval_strategy='epoch',
    save_strategy='epoch',
    load_best_model_at_end=True,
    metric_for_best_model='eval_loss',
    report_to='wandb',
    max_seq_length=128,
    dataset_text_field='text',
)

trainer = SFTTrainer(
    model=model,
    args=cfg,
    train_dataset=ds_train,
    eval_dataset=ds_val,
    tokenizer=tokenizer,
)
trainer.train()

/usr/local/lib/python3.12/dist-packages/bitsandbytes/backends/cuda/ops.py:213: FutureWarning: _check_is_size will be removed in a future PyTorch release along with guard_size_oblivious.     Use _check(i >= 0) instead.
  torch._check_is_size(blocksize)


trainable params: 425,984 || all params: 1,236,240,384 || trainable%: 0.0345


Map:   0%|          | 0/20135 [00:00<?, ? examples/s]

Map:   0%|          | 0/2517 [00:00<?, ? examples/s]

/usr/local/lib/python3.12/dist-packages/trl/trainer/sft_trainer.py:401: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `SFTTrainer.__init__`. Use `processing_class` instead.
  super().__init__(
wandb: WARNING The `run_name` is currently set to the same value as `TrainingArguments.output_dir`. If this was not intended, please specify a different run name by setting the `TrainingArguments.run_name` parameter.
/usr/local/lib/python3.12/dist-packages/torch/_dynamo/eval_frame.py:1263: UserWarning: torch.utils.checkpoint: the use_reentrant parameter should be passed explicitly. Starting in PyTorch 2.9, calling checkpoint without use_reentrant will raise an exception. use_reentrant=False is recommended, but if you need to preserve the current default behavior, you can pass use_reentrant=True. Refer to docs for more details on the differences between the two variants.
  return fn(*args, **kwargs)


Epoch,Training Loss,Validation Loss


Epoch,Training Loss,Validation Loss
1,4.457000,2.220682
2,4.317300,2.193269


/usr/local/lib/python3.12/dist-packages/torch/_dynamo/eval_frame.py:1263: UserWarning: torch.utils.checkpoint: the use_reentrant parameter should be passed explicitly. Starting in PyTorch 2.9, calling checkpoint without use_reentrant will raise an exception. use_reentrant=False is recommended, but if you need to preserve the current default behavior, you can pass use_reentrant=True. Refer to docs for more details on the differences between the two variants.
  return fn(*args, **kwargs)


TrainOutput(global_step=5034, training_loss=4.5273598802435435, metrics={'train_runtime': 18218.0296, 'train_samples_per_second': 2.21, 'train_steps_per_second': 0.276, 'total_flos': 2.4344368869187584e+16, 'train_loss': 4.5273598802435435, 'epoch': 2.0})

In [ ]:
adapter_path = './adapters/exp1_attn_r4'
trainer.model.save_pretrained(adapter_path)
tokenizer.save_pretrained(adapter_path)

art = wandb.Artifact('exp1_attn_r4', type='model')
art.add_dir(adapter_path)
run.log_artifact(art)

wandb: Adding directory to artifact (adapters/exp1_attn_r4)... Done. 0.1s


<Artifact exp1_attn_r4>

## Оцениваем результаты

In [ ]:
val_loss = trainer.evaluate()['eval_loss']
perplexity = round(float(np.exp(val_loss)), 2)

model = PeftModel.from_pretrained(load_model(), adapter_path)
model.eval()

metrics = compute_metrics(model, df_test)
metrics['perplexity'] = perplexity

wandb.log({f'test/{k}': v for k, v in metrics.items()})
run.finish()

print('base:', base_metrics)
print('finetuned:', metrics)

/usr/local/lib/python3.12/dist-packages/bitsandbytes/backends/cuda/ops.py:213: FutureWarning: _check_is_size will be removed in a future PyTorch release along with guard_size_oblivious.     Use _check(i >= 0) instead.
  torch._check_is_size(blocksize)
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
/usr/local/lib/python3.12/dist-packages/bitsandbytes/backends/cuda/ops.py:468: FutureWarning: _check_is_size will be removed in a future PyTorch release along with guard_size_oblivious.     Use _check(i >= 0) instead.
  torch._check_is_size(blocksize)
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:128001 for open-end gene

eval/loss,█▁▁
eval/runtime,▁▆█
eval/samples_per_second,█▃▁
eval/steps_per_second,█▅▁
test/bertscore,▁
test/distinct1,▁
test/distinct2,▁
test/perplexity,▁
test/rouge1,▁
test/rougeL,▁
+5,...


base: {'rouge1': np.float64(0.1103), 'rougeL': np.float64(0.0789), 'bertscore': 0.8269, 'distinct1': 0.3322, 'distinct2': 0.7304}
finetuned: {'rouge1': np.float64(0.1091), 'rougeL': np.float64(0.0875), 'bertscore': 0.8021, 'distinct1': 0.118, 'distinct2': 0.3115, 'perplexity': 8.96}


Итоги эксперимента 1
ROUGE-L вырос с 0.0789 до 0.0875 — модель стала генерить ответы структурно ближе к референсным.  

Для r=4 только на attention-слоях результат нормальный.
Distinct-1/2 упал (0.33 → 0.12, 0.73 → 0.31) — это именно то что хотели: модель перестала "фантазировать" и начала отвечать в стиле саппорта.  

Разнообразие упало потому что домен узкий, и это ожидаемо.
BERTScore 0.80 — семантически ответы всё ещё близки к референсу, просто модель стала более лаконичной и по делу.  
Perplexity = 8.96 после 2 эпох — модель хорошо освоила распределение текстов поддержки.  
Вывод: файнтюн сработал — модель адаптировалась под стиль customer support.  

ROUGE-L подрос, perplexity низкая, distinct упал в нужную сторону.  

Лёгкий LoRA на attention с r=4 оказался достаточным для доменной адаптации.

## Чат с моделью

In [ ]:
history = []

@torch.no_grad()
def chat_respond():
    tokens = tokenizer.apply_chat_template(
        history, return_tensors='pt', add_generation_prompt=True
    ).to(model.device)
    generated = model.generate(
        tokens,
        max_new_tokens=200,
        temperature=0.7,
        top_p=0.9,
        do_sample=True
    )
    response = tokenizer.decode(generated[0][tokens.shape[1]:], skip_special_tokens=True)
    return response

def on_send(b):
    user_msg = inp.value.strip()
    if not user_msg:
        return
    inp.value = ''
    history.append({'role': 'user', 'content': user_msg})
    with chat_out:
        print(f'Пользователь: {user_msg}')
    model_reply = chat_respond()
    history.append({'role': 'assistant', 'content': model_reply})
    with chat_out:
        print(f'Модель: {model_reply}')
        print()

def on_clear(b):
    history.clear()
    chat_out.clear_output()

inp = widgets.Text(
    placeholder='введите сообщение...',
    layout=widgets.Layout(width='70%')
)
btn = widgets.Button(description='Отправить', button_style='primary')
clr = widgets.Button(description='Очистить', button_style='warning')

chat_out = widgets.Output()

btn.on_click(on_send)
clr.on_click(on_clear)
display(widgets.VBox([widgets.HBox([inp, btn, clr]), chat_out]))